File: client/src/main/scala/stasis/client/ops/scheduling/DefaultOperationExecutor.scala
package stasis.client.ops.scheduling

import java.nio.file.Path
import java.time.Instant

import scala.concurrent.ExecutionContext
import scala.concurrent.Future
import scala.util.control.NonFatal

import org.apache.pekko.Done
import org.apache.pekko.actor.typed.ActorSystem
import org.apache.pekko.actor.typed.scaladsl.LoggerOps
import org.apache.pekko.util.Timeout
import org.slf4j.Logger
import org.slf4j.LoggerFactory

import stasis.client.collection.rules.Specification
import stasis.client.encryption.secrets.DeviceSecret
import stasis.client.ops.ParallelismConfig
import stasis.client.ops.backup
import stasis.client.ops.exceptions.OperationExecutionFailure
import stasis.client.ops.exceptions.OperationStopped
import stasis.client.ops.recovery
import stasis.client.tracking.state.OperationState
import stasis.layers.persistence.memory.MemoryStore
import stasis.layers.telemetry.TelemetryContext
import stasis.shared.model.datasets.DatasetDefinition
import stasis.shared.model.datasets.DatasetEntry
import stasis.shared.ops.Operation

class DefaultOperationExecutor(
  config: DefaultOperationExecutor.Config,
  secret: DeviceSecret
)(implicit
  system: ActorSystem[Nothing],
  telemetry: TelemetryContext,
  parallelismConfig: ParallelismConfig,
  timeout: Timeout,
  backupProviders: backup.Providers,
  recoveryProviders: recovery.Providers
) extends OperationExecutor {

  private implicit val ec: ExecutionContext = system.executionContext

  private val log: Logger = LoggerFactory.getLogger(this.getClass.getName)

  private val activeOperations: MemoryStore[Operation.Id, Operation] =
    MemoryStore(name = "executor-active-operations-store")

  private val completedOperations: MemoryStore[Operation.Id, Operation.Type] =
    MemoryStore(name = "executor-completed-operations-store")

  override def active: Future[Map[Operation.Id, Operation.Type]] =
    activeOperations.entries.map(_.map { case (id, operation) => (id, operation.`type`) })

  override def completed: Future[Map[Operation.Id, Operation.Type]] =
    completedOperations.entries

  override def rules: Future[Specification] =
    SchedulingConfig.rules(file = config.backup.rulesFile).flatMap(Specification.untracked)

  override def startBackupWithRules(
    definition: DatasetDefinition.Id
  ): Future[Operation.Id] =
    for {
      _ <- requireUniqueOperation(ofType = Operation.Type.Backup)
      rules <- SchedulingConfig.rules(file = config.backup.rulesFile)
      descriptor <- backup.Backup.Descriptor(
        definition = definition,
        collector = backup.Backup.Descriptor.Collector.WithRules(rules),
        deviceSecret = secret,
        limits = config.backup.limits
      )
      operation = backup.Backup(descriptor = descriptor)
      _ <- activeOperations.put(operation.id, operation)
    } yield {
      operation.run()
      operation.id
    }

  override def startBackupWithEntities(
    definition: DatasetDefinition.Id,
    entities: Seq[Path]
  ): Future[Operation.Id] =
    for {
      _ <- requireUniqueOperation(ofType = Operation.Type.Backup)
      descriptor <- backup.Backup.Descriptor(
        definition = definition,
        collector = backup.Backup.Descriptor.Collector.WithEntities(entities = entities),
        deviceSecret = secret,
        limits = config.backup.limits
      )
      operation = backup.Backup(descriptor = descriptor)
      _ <- activeOperations.put(operation.id, operation)
    } yield {
      operation.run()
      operation.id
    }

  override def resumeBackup(
    operation: Operation.Id
  ): Future[Operation.Id] =
    for {
      _ <- requireUniqueOperation(ofType = Operation.Type.Backup)
      state <- backupProviders.track.state.map(_.get(operation))
      state <- requireValidOperationState(operation = operation, state = state)
      descriptor <- backup.Backup.Descriptor(
        definition = state.definition,
        collector = backup.Backup.Descriptor.Collector.WithState(state = state),
        deviceSecret = secret,
        limits = config.backup.limits
      )
      operation = backup.Backup(descriptor = descriptor)
      _ <- activeOperations.put(operation.id, operation)
    } yield {
      operation.run()
      operation.id
    }

  override def startRecoveryWithDefinition(
    definition: DatasetDefinition.Id,
    until: Option[Instant],
    query: Option[recovery.Recovery.PathQuery],
    destination: Option[recovery.Recovery.Destination]
  ): Future[Operation.Id] =
    for {
      _ <- requireUniqueOperation(ofType = Operation.Type.Recovery)
      descriptor <- recovery.Recovery.Descriptor(
        collector = recovery.Recovery.Descriptor.Collector.WithDefinition(definition, until),
        query = query,
        destination = destination,
        deviceSecret = secret
      )
      operation = recovery.Recovery(descriptor = descriptor)
      _ <- activeOperations.put(operation.id, operation)
    } yield {
      operation.run()
      operation.id
    }

  override def startRecoveryWithEntry(
    entry: DatasetEntry.Id,
    query: Option[recovery.Recovery.PathQuery],
    destination: Option[recovery.Recovery.Destination]
  ): Future[Operation.Id] =
    for {
      _ <- requireUniqueOperation(ofType = Operation.Type.Recovery)
      descriptor <- recovery.Recovery.Descriptor(
        query = query,
        destination = destination,
        collector = recovery.Recovery.Descriptor.Collector.WithEntry(entry),
        deviceSecret = secret
      )
      operation = recovery.Recovery(descriptor = descriptor)
      _ <- activeOperations.put(operation.id, operation)
    } yield {
      operation.run()
      operation.id
    }

  override def startExpiration(): Future[Operation.Id] =
    Future.failed(new NotImplementedError("Expiration is not supported")) // TODO - implement

  override def startValidation(): Future[Operation.Id] =
    Future.failed(new NotImplementedError("Validation is not supported")) // TODO - implement

  override def startKeyRotation(): Future[Operation.Id] =
    Future.failed(new NotImplementedError("Key rotation is not supported")) // TODO - implement

  override def stop(operation: Operation.Id): Future[Option[Done]] =
    activeOperations
      .get(operation)
      .map(_.map { operation =>
        operation.stop()
        Done
      })

  private def requireUniqueOperation(ofType: Operation.Type): Future[Done] =
    active.flatMap { active =>
      active.find(_._2 == ofType) match {
        case Some((operation, _)) =>
          val opType = ofType.toString
          val message = s"Cannot start [$opType] operation; [$opType] with ID [${operation.toString}] is already active"
          log.errorN(message)
          Future.failed(new OperationExecutionFailure(message))

        case None =>
          Future.successful(Done)
      }
    }

  private def requireValidOperationState[S <: OperationState](
    operation: Operation.Id,
    state: Option[S]
  ): Future[S] =
    state match {
      case Some(state) if !state.isCompleted =>
        Future.successful(state)

      case Some(_) =>
        val message = s"Cannot resume operation with ID [${operation.toString}]; operation already completed"
        log.errorN(message)
        Future.failed(new OperationExecutionFailure(message))

      case None =>
        val message = s"Cannot resume operation with ID [${operation.toString}]; no existing state was found"
        log.errorN(message)
        Future.failed(new OperationExecutionFailure(message))
    }

  implicit class RunnableOperation(operation: Operation) {
    def run(): Unit = {
      val _ = runWithResult()
    }

    def runWithResult(): Future[Done] = {
      log.debugN("Starting operation [{}]", operation.id)
      operation
        .start()
        .recover {
          case NonFatal(e: OperationStopped) =>
            log.debugN(e.message)
            Done

          case NonFatal(e) =>
            log.errorN(
              "Failure encountered when running operation [{}]: [{} - {}]",
              operation.id,
              e.getClass.getSimpleName,
              e.getMessage
            )
            Done
        }
        .flatMap { _ =>
          for {
            _ <- activeOperations.delete(operation.id)
            _ <- completedOperations.put(operation.id, operation.`type`)
          } yield {
            Done
          }
        }
    }
  }
}

object DefaultOperationExecutor {
  final case class Config(
    backup: Config.Backup
  )

  object Config {
    final case class Backup(
      rulesFile: Path,
      limits: backup.Backup.Limits
    )
  }
}

File: client-android/lib/src/main/kotlin/stasis/client_android/lib/ops/scheduling/DefaultOperationExecutor.kt
package stasis.client_android.lib.ops.scheduling

import stasis.client_android.lib.ops.backup.Backup as BackupOp
import stasis.client_android.lib.ops.backup.Providers as BackupProviders
import stasis.client_android.lib.ops.recovery.Providers as RecoveryProviders
import stasis.client_android.lib.ops.recovery.Recovery as RecoveryOp
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.withContext
import stasis.client_android.lib.collection.rules.Rule
import stasis.client_android.lib.encryption.secrets.DeviceSecret
import stasis.client_android.lib.model.server.datasets.DatasetDefinitionId
import stasis.client_android.lib.model.server.datasets.DatasetEntryId
import stasis.client_android.lib.ops.Operation
import stasis.client_android.lib.ops.OperationId
import stasis.client_android.lib.tracking.state.OperationState
import stasis.client_android.lib.utils.Try.Failure
import stasis.client_android.lib.utils.Try.Success
import java.nio.file.Path
import java.time.Instant
import java.util.concurrent.ConcurrentHashMap

class DefaultOperationExecutor(
    private val config: Config,
    private val deviceSecret: () -> DeviceSecret,
    private val backupProviders: BackupProviders,
    private val recoveryProviders: RecoveryProviders,
    operationDispatcher: CoroutineDispatcher
) : OperationExecutor {
    private val operationScope = CoroutineScope(operationDispatcher)

    private val activeOperations: ConcurrentHashMap<OperationId, Operation> =
        ConcurrentHashMap()

    private val completeOperations: ConcurrentHashMap<OperationId, Operation.Type> =
        ConcurrentHashMap()

    override suspend fun active(): Map<OperationId, Operation.Type> =
        activeOperations.mapValues { it.value.type() }

    override suspend fun completed(): Map<OperationId, Operation.Type> =
        completeOperations

    override suspend fun find(operation: OperationId): Operation.Type? =
        when (val active = activeOperations[operation]) {
            null -> completeOperations[operation]
            else -> active.type()
        }

    override suspend fun startBackupWithRules(
        definition: DatasetDefinitionId,
        rules: List<Rule>,
        f: (Throwable?) -> Unit
    ): OperationId = asUniqueOperation(ofType = Operation.Type.Backup, callback = f) {
        val descriptor = BackupOp.Descriptor(
            definition = definition,
            collector = BackupOp.Descriptor.Collector.WithRules(rules),
            deviceSecret = deviceSecret(),
            limits = config.backup.limits,
            providers = backupProviders
        )

        val operationId = when (descriptor) {
            is Success -> {
                val operation = BackupOp(descriptor.value, backupProviders)

                activeOperations[operation.id] = operation

                operation.start(withScope = operationScope) {
                    activeOperations.remove(operation.id)
                    completeOperations[operation.id] = operation.type()
                    f(it)
                }

                operation.id
            }

            is Failure -> {
                f(descriptor.exception)
                Operation.generateId()
            }
        }

        operationId
    }

    override suspend fun startBackupWithEntities(
        definition: DatasetDefinitionId,
        entities: List<Path>,
        f: (Throwable?) -> Unit
    ): OperationId = asUniqueOperation(ofType = Operation.Type.Backup, callback = f) {
        val descriptor = BackupOp.Descriptor(
            definition = definition,
            collector = BackupOp.Descriptor.Collector.WithEntities(entities = entities),
            deviceSecret = deviceSecret(),
            limits = config.backup.limits,
            providers = backupProviders
        )

        val operationId = when (descriptor) {
            is Success -> {
                val operation = BackupOp(descriptor.value, backupProviders)

                activeOperations[operation.id] = operation

                operation.start(withScope = operationScope) {
                    activeOperations.remove(operation.id)
                    completeOperations[operation.id] = operation.type()
                    f(it)
                }

                operation.id
            }

            is Failure -> {
                f(descriptor.exception)
                Operation.generateId()
            }
        }

        operationId
    }

    override suspend fun resumeBackup(
        operation: OperationId,
        f: (Throwable?) -> Unit
    ): OperationId = asUniqueOperation(ofType = Operation.Type.Backup, callback = f) {
        asValidOperationState(
            operation = operation,
            state = backupProviders.track.stateOf(operation),
            callback = f
        ) { state ->
            val descriptor = BackupOp.Descriptor(
                definition = state.definition,
                collector = BackupOp.Descriptor.Collector.WithState(state = state),
                deviceSecret = deviceSecret(),
                limits = config.backup.limits,
                providers = backupProviders
            )

            when (descriptor) {
                is Success -> {
                    val actualOperation = BackupOp(descriptor.value, backupProviders)

                    activeOperations[actualOperation.id] = actualOperation

                    actualOperation.start(withScope = operationScope) {
                        activeOperations.remove(actualOperation.id)
                        completeOperations[actualOperation.id] = actualOperation.type()
                        f(it)
                    }

                }

                is Failure -> {
                    f(descriptor.exception)
                }
            }

            operation
        }
    }

    override suspend fun startRecoveryWithDefinition(
        definition: DatasetDefinitionId,
        until: Instant?,
        query: RecoveryOp.PathQuery?,
        destination: RecoveryOp.Destination?,
        f: (Throwable?) -> Unit
    ): OperationId = asUniqueOperation(ofType = Operation.Type.Recovery, callback = f) {
        val descriptor = RecoveryOp.Descriptor(
            collector = RecoveryOp.Descriptor.Collector.WithDefinition(definition, until),
            query = query,
            destination = destination,
            deviceSecret = deviceSecret(),
            providers = recoveryProviders
        )

        val operationId = when (descriptor) {
            is Success -> {
                val operation = RecoveryOp(descriptor.value, recoveryProviders)

                activeOperations[operation.id] = operation

                operation.start(withScope = operationScope) {
                    activeOperations.remove(operation.id)
                    completeOperations[operation.id] = operation.type()
                    f(it)
                }

                operation.id
            }

            is Failure -> {
                f(descriptor.exception)
                Operation.generateId()
            }
        }

        operationId
    }

    override suspend fun startRecoveryWithEntry(
        entry: DatasetEntryId,
        query: RecoveryOp.PathQuery?,
        destination: RecoveryOp.Destination?,
        f: (Throwable?) -> Unit
    ): OperationId = asUniqueOperation(ofType = Operation.Type.Recovery, callback = f) {
        val descriptor = RecoveryOp.Descriptor(
            collector = RecoveryOp.Descriptor.Collector.WithEntry(entry),
            query = query,
            destination = destination,
            deviceSecret = deviceSecret(),
            providers = recoveryProviders
        )

        val operationId = when (descriptor) {
            is Success -> {
                val operation = RecoveryOp(descriptor.value, recoveryProviders)

                activeOperations[operation.id] = operation

                operation.start(withScope = operationScope) {
                    activeOperations.remove(operation.id)
                    completeOperations[operation.id] = operation.type()
                    f(it)
                }

                operation.id
            }

            is Failure -> {
                f(descriptor.exception)
                Operation.generateId()
            }
        }

        operationId
    }

    override suspend fun startExpiration(f: (Throwable?) -> Unit): OperationId {
        throw NotImplementedError("Expiration is not supported") // TODO - implement
    }

    override suspend fun startValidation(f: (Throwable?) -> Unit): OperationId {
        throw NotImplementedError("Validation is not supported") // TODO - implement
    }

    override suspend fun startKeyRotation(f: (Throwable?) -> Unit): OperationId {
        throw NotImplementedError("Key rotation is not supported") // TODO - implement
    }

    override suspend fun stop(operation: OperationId) {
        when (val activeOperation = activeOperations[operation]) {
            null -> throw IllegalArgumentException("Failed to stop [$operation]; operation not found")
            else -> activeOperation.stop()
        }
    }

    private suspend fun asUniqueOperation(
        ofType: Operation.Type,
        callback: (Throwable?) -> Unit,
        f: suspend () -> OperationId
    ): OperationId = withContext(operationScope.coroutineContext) {
        when (val operation = active().filterValues { it == ofType }.toList().firstOrNull()?.first) {
            null -> f() // operation doesn't exist; can proceed
            else -> {
                callback(
                    IllegalArgumentException(
                        "Cannot start [$ofType] operation; [$ofType] with ID [$operation] is already active"
                    )
                )

                operation
            }
        }
    }

    private suspend fun <T : OperationState> asValidOperationState(
        operation: OperationId,
        state: T?,
        callback: (Throwable?) -> Unit,
        f: suspend (T) -> OperationId
    ): OperationId = withContext(operationScope.coroutineContext) {
        when {
            state != null && state.completed == null -> f(state)
            state != null -> callback(
                IllegalArgumentException("Cannot resume operation with ID [$operation]; operation already completed")
            )
            else -> callback(
                IllegalArgumentException("Cannot resume operation with ID [$operation]; no existing state was found")
            )
        }

        operation
    }

    data class Config(
        val backup: Backup
    ) {
        data class Backup(
            val limits: BackupOp.Descriptor.Limits
        )
    }
}
